### PR TITLE
Fixed #33589 -- Made admin calendar widget escape all char occurrences in date input formats.

### PR DESCRIPTION
**Pull request by ankurr0y**
Now the replacement of the string of characters is for all occurrences and not just the first occurrence.

### CODE REVIEW

**Review by David-Wobrock — COMMENTED** [review]
Hi @ankurr0y , thanks for the patch!
Could you add a regression test please? :)

See Mariusz' comment https://code.djangoproject.com/ticket/33589#comment:2 (as the previously suggested patch was closed due to inactivity and was also missing a unit test https://github.com/django/django/pull/15460)

**Comment by ankurr0y** [comment]
All right, I added a unit test which covers all of the expressions and replaces them successfully. However, there is a problem on a different unit test. On the unit test 'init', the callback function has an error for a unexpected \.
Screenshot from 2022-06-04 09-52-08

This error occurs with both the old and the new code. I have separated the \\ regex from the others for now.

File: django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js [hunk 1]
```diff
@@ -389,11 +389,13 @@
...
-                .replace('\n', '\\n')
-                .replace('\t', '\\t')
-                .replace("'", "\\'");
+            // this \\ doesn't work with replace or replaceall, results in error thrown by init test
+            format = format.replace(/\\/g, "\\\\");
+
+            format = format.replace(new RegExp('\r', 'g'), '\\r')
+                .replace(new RegExp('\n', 'g'), '\\n')
+                .replace(new RegExp('\t', 'g'), '\\t')
+                .replace(new RegExp("'", 'g'), "\\'");
```
Comment (David-Wobrock) [inline django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:398]: We can keep one expression as it was before, no need to reassign `format`
```suggestion
            format = format.replace(/\\/g, "\\\\")
                .replace(new RegExp('\r', 'g'), '\\r')
                .replace(new RegExp('\n', 'g'), '\\n')
                .replace(new RegExp('\t', 'g'), '\\t')
                .replace(new RegExp("'", 'g'), "\\'");
```

File: js_tests/admin/DateTimeShortcuts.test.js [hunk 2]
```diff
@@ -40,3 +40,15 @@ QUnit.test('time zone offset warning', function(assert) {
...
+
+QUnit.test('Calendar regular expression callback', function(assert) {
+    const $ = django.jQuery;
+    let format = "\n February \n 12 \n 1998 \r \r \r \t \t \t \\ \\ ' '";
+    format = format.replace(/\\/g, "\\\\");
+    format = format.replace(new RegExp('\r', 'g'), '\\r')
+            .replace(new RegExp('\n', 'g'), '\\n')
+            .replace(new RegExp('\t', 'g'), '\\t')
+            .replace(new RegExp("'", 'g'), "\\'")
+            ;
```
Comment (David-Wobrock) [inline js_tests/admin/DateTimeShortcuts.test.js:52]: We shouldn't duplicate the code from the actual function, but instead be able to use the code from `django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js`

We could add a function that will specifically escape the date format and call it here:
```
const format = "\n %Y \r %m \r \t %d \t %H \\ %M \\ '%S'\n";
const escapedFormat = DateTimeShortcuts.escapeFormat(format);
assert.equal(escapedFormat, "\\n %Y \\r %m \\r \\t %d \\t %H \\\\ %M \\\\ \\'%S\\'\n");
```
Something in this vein.

Comment (ankurr0y) [reply js_tests/admin/DateTimeShortcuts.test.js:52]: Ok, thank you I will do that.

Comment (ankurr0y) [reply js_tests/admin/DateTimeShortcuts.test.js]: I have done as you suggested, escapeFormat is now a separate function.

File: js_tests/admin/DateTimeShortcuts.test.js [hunk 3]
```diff
@@ -40,3 +40,10 @@ QUnit.test('time zone offset warning', function(assert) {
     $('body').attr('data-admin-utc-offset', savedOffset);
     assert.equal($('.timezonewarning').text(), 'Note: You are 1 hour behind server time.');
 });
+
+QUnit.test('Calendar regular expression callback', function(assert) {
+    const $ = django.jQuery;
```
Comment (David-Wobrock) [inline js_tests/admin/DateTimeShortcuts.test.js:45]: Is not used
```suggestion
```

**Review by David-Wobrock — APPROVED** [review]

**Review by David-Wobrock — APPROVED** [review]
Thanks for the changes @ankurr0y!
I'll let a maintainer have another look and edit/merge the patch when ready :)

File: django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js [hunk 4]
```diff
@@ -389,11 +389,8 @@
         handleCalendarCallback: function(num) {
             let format = get_format('DATE_INPUT_FORMATS')[0];
             // the format needs to be escaped a little
-            format = format.replace('\\', '\\\\')
-                .replace('\r', '\\r')
-                .replace('\n', '\\n')
-                .replace('\t', '\\t')
-                .replace("'", "\\'");
+
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:392]: Chop blank line.

File: django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js [hunk 5]
```diff
@@ -406,6 +403,14 @@
...
             DateTimeShortcuts.calendarInputs[num].focus();
             DateTimeShortcuts.dismissCalendar(num);
+        },
+        escapeFormat: function(format) {
+            format = format.replace(/\\/g, "\\\\")
+                .replace(new RegExp('\r', 'g'), '\\r')
+                .replace(new RegExp('\n', 'g'), '\\n')
+                .replace(new RegExp('\t', 'g'), '\\t')
+                .replace(new RegExp("'", 'g'), "\\'");
+            return format
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:413]:
```suggestion
            return format;
```

File: django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js [hunk 6]
```diff
@@ -406,6 +403,14 @@
             DateTimeShortcuts.calendarInputs[num].value = d.strftime(get_format('DATE_INPUT_FORMATS')[0]);
             DateTimeShortcuts.calendarInputs[num].focus();
             DateTimeShortcuts.dismissCalendar(num);
+        },
+        escapeFormat: function(format) {
+            format = format.replace(/\\/g, "\\\\")
+                .replace(new RegExp('\r', 'g'), '\\r')
+                .replace(new RegExp('\n', 'g'), '\\n')
+                .replace(new RegExp('\t', 'g'), '\\t')
+                .replace(new RegExp("'", 'g'), "\\'");
```
Comment (felixxm) [inline django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:412]: Do we need to create a new `RegExp` objects? `/\t/g` syntax works for me:
```suggestion
            format = format.replace(/\\/g, "\\\\")
                .replace(/\r/g, "\\r")
                .replace(/\n/g, "\\n")
                .replace(/\t/g, "\\t")
                .replace(/'/g, "\\'");
```

File: js_tests/admin/DateTimeShortcuts.test.js [hunk 7]
```diff
@@ -40,3 +40,9 @@ QUnit.test('time zone offset warning', function(assert) {
     $('body').attr('data-admin-utc-offset', savedOffset);
     assert.equal($('.timezonewarning').text(), 'Note: You are 1 hour behind server time.');
 });
+
+QUnit.test('Calendar regular expression callback', function(assert) {
```
Comment (felixxm) [inline js_tests/admin/DateTimeShortcuts.test.js:44]:
```suggestion
QUnit.test('calendar escape format', function(assert) {
```

File: js_tests/admin/DateTimeShortcuts.test.js [hunk 8]
```diff
@@ -40,3 +40,9 @@ QUnit.test('time zone offset warning', function(assert) {
...
+    let format = "\n February \n 12 \n 1998 \r \r \r \t \t \t \\ \\ ' '";
```
Comment (felixxm) [inline js_tests/admin/DateTimeShortcuts.test.js:45]: This is not really a format. I'd use sth more realistic:
```suggestion
    let date_input_format = "'%d \r\t\\\n %B \n\\\r\t %Y'";
```

**Review by felixxm — COMMENTED** [review]
@ankurr0y Thanks for this patch :+1:

File: django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js (hunk 4 above)
Comment (ankurr0y) [reply django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js:392]: Thanks man, I have committed your suggestions.

File: js_tests/admin/DateTimeShortcuts.test.js [hunk 9]
```diff
@@ -40,3 +40,9 @@ QUnit.test('time zone offset warning', function(assert) {
     $('body').attr('data-admin-utc-offset', savedOffset);
     assert.equal($('.timezonewarning').text(), 'Note: You are 1 hour behind server time.');
 });
+
+QUnit.test('calendar escape format', function(assert) {
+    const date_input_format = "'%d \r\t\\\n %B \n\\\r\t %Y'";
```
Comment (David-Wobrock) [inline js_tests/admin/DateTimeShortcuts.test.js:45]: The case should be something like `dateInputFormat`

File: js_tests/admin/DateTimeShortcuts.test.js [hunk 10]
```diff
@@ -40,3 +40,9 @@ QUnit.test('time zone offset warning', function(assert) {
...
+    format = DateTimeShortcuts.escapeFormat(date_input_format);
+    assert.equal(format, "\\n February \\n 12 \\n 1998 \\r \\r \\r \\t \\t \\t \\\\ \\\\ \\' \\'");
```
Comment (David-Wobrock) [inline js_tests/admin/DateTimeShortcuts.test.js:47]: Why are the tests green? Shouldn't this fail as we didn't update the expected string 🤔

Comment (felixxm) [reply js_tests/admin/DateTimeShortcuts.test.js:47]: It fails as expected. You're a first-time contributor so we need to manually approve each run.

Comment (felixxm) [inline js_tests/admin/DateTimeShortcuts.test.js:47]:
```suggestion
    let dateInputFormat = "'%d \r\t\\\n %B \n\\\r\t %Y'";
    dateInputFormat = DateTimeShortcuts.escapeFormat(dateInputFormat);
    assert.equal(dateInputFormat, "\\'%d \\r\\t\\\\\\n %B \\n\\\\\\r\\t %Y\\'");
```

**Review by David-Wobrock — APPROVED** [review]

**Comment by felixxm** [comment]
@ankurr0y Thanks :+1: Welcome aboard :boat:

**Comment by felixxm** [comment]
Closing per [comment](https://code.djangoproject.com/ticket/33589#comment:13).

@ankurr0y Would you like to prepare a patch for ticket-33784?

**Pull request closed without merging**